posthog/temporal/llm_analytics/trace_summarization/embedding.py

```diff
@@ -0,0 +1,150 @@
+from posthog.clickhouse.client.execute import sync_execute
+from posthog.models.team import Team
+from posthog.sync import database_sync_to_async
+from posthog.temporal.llm_analytics.trace_summarization import constants
+from posthog.temporal.llm_analytics.trace_summarization.models import (
+    EmbeddingActivityResult,
+    SingleEmbeddingResult,
+    SummaryRow,
+)
+
+from ee.hogai.llm_traces_summaries.constants import LLM_TRACES_SUMMARIES_DOCUMENT_TYPE, LLM_TRACES_SUMMARIES_PRODUCT
+from ee.hogai.llm_traces_summaries.tools.embed_summaries import LLMTracesSummarizerEmbedder
+
+logger = structlog.get_logger(__name__)
+
+
+@temporalio.activity.defn
+async def embed_summaries_activity(
+    trace_ids: list[str],
+    team_id: int,
+    mode: str,
+    workflow_start_time: str | None = None,
+) -> EmbeddingActivityResult:
+    """
+    Fetch summaries from ClickHouse and queue embeddings via Kafka.
+
+    Uses workflow_start_time to efficiently filter events to a narrow time window.
+    """
+
+    def _fetch_summaries_and_setup(
+        trace_ids: list[str], team_id: int, mode: str, workflow_start_time: str | None = None
+    ) -> tuple[list[SummaryRow], LLMTracesSummarizerEmbedder, str]:
+        """Fetch summaries from ClickHouse and setup embedder (sync operations)."""
+        timestamp_filter = ""
+        if workflow_start_time:
+            timestamp_filter = "AND timestamp >= %(time_from)s AND timestamp <= %(time_to)s"
+
+        # Fetch summaries from ClickHouse events
+        # Always filter for 'detailed' mode to get richest embeddings
+        query = f"""
+            SELECT
+                JSONExtractString(properties, '$ai_trace_id') as trace_id,
+                JSONExtractString(properties, '$ai_summary_title') as title,
+                JSONExtractString(properties, '$ai_summary_flow_diagram') as flow_diagram,
+                JSONExtractString(properties, '$ai_summary_bullets') as bullets,
+                JSONExtractString(properties, '$ai_summary_interesting_notes') as notes
+            FROM events
+            WHERE team_id = %(team_id)s
+                AND event = %(event_name)s
+                AND JSONExtractString(properties, '$ai_trace_id') IN %(trace_ids)s
+                AND JSONExtractString(properties, '$ai_summary_mode') = %(summary_mode)s
+                {timestamp_filter}
+            ORDER BY timestamp DESC
+        """
+
+        params = {
+            "team_id": team_id,
+            "event_name": constants.EVENT_NAME_TRACE_SUMMARY,
+            "trace_ids": trace_ids,
+            "summary_mode": constants.DEFAULT_MODE,  # Always use 'detailed' for richest embeddings
+        }
+
+        # Add timestamp params if filter is active
+        # Use strftime to produce ClickHouse-compatible format (no timezone offset)
+        # Window covers workflow start minus buffer, to workflow start plus max execution time plus buffer
+        if workflow_start_time:
```

Comment:
This is _probably_ fine, but I wonder if it will ever lead to some edge case bugs. I also wonder if there's a way to know which timestamp start and end to put as we enter this activity 🤔 We might not be able to, and this might be the best way to do it. We could probably event increase the time range without a significant performance hit, since we're filtering by the summarization event names.